[JENKINS-56349] Protect against null pointer exception

proski opened the pull request:
```
*  [JENKINS-56349] Make StashPullRequestsBuilder constructor take two arguments
   
   Call that constructor from StashBuildTrigger#start() instead of using
   setters. Check that project is not null before passing it to the
   StashPullRequestsBuilder constructor.
   
   Remove setters from StashPullRequestsBuilder, they are not used anymore.

*  [JENKINS-56349] Protect StashBuildTrigger#run() against crash if called early
   
   Trigger#run() can be called before Trigger#start() according to
   https://javadoc.jenkins.io/hudson/triggers/Trigger.html#run--
   
   Improve the log message if the project is disabled. Lower the level to
   "fine", it should be clear why the disabled project is not building pull
   requests.
   
   Don't log anything when starting the build, the builder logs a similar
   message.
   
   Make stop() set the builder to null to revert the effect of start().
```
Please see:
https://issues.jenkins-ci.org/browse/JENKINS-56349

Comment by Eyal Goren:
https://wiki.jenkins.io/display/JENKINS/Stash+pullrequest+builder+plugin

This is not a fix for pipeline support. This is a fix for a crash found when trying to setup a pipeline.

The first commit fixes the abomination that amounts to constructing an object from outside the class. The second commit adds some defensive measures to `StashBuildTrigger#run()` if it's run too early.

proski committed:
[JENKINS-56349] Protect StashBuildTrigger#run() against crash if called early

Trigger#run() can be called before Trigger#start() according to
https://javadoc.jenkins.io/hudson/triggers/Trigger.html#run--

Improve the log message if the project is disabled. Lower the level to
"fine", it should be clear why the disabled project is not building pull
requests.

Don't log anything when starting the build, the builder logs a similar
message.

Make stop() set the builder to null to revert the effect of start().

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashPullRequestsBuilder.java:
```diff
@@ -18,12 +18,15 @@
-        return new StashPullRequestsBuilder();
-    }
+    public StashPullRequestsBuilder(AbstractProject project, StashBuildTrigger trigger) {
+        if (project == null || trigger == null) {
```
Can you change it to `Objects.requireNonNull()`?

proski replied:
Sure, good idea. The trigger argument can be annotated as non-null, it's always called with `this`.

In the long term, we should probably eliminate StashPullRequiestBuilder, it's a thin "appendix" to StashBuildTrigger with 1-1 relationship. But that would be another PR.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.java:
```diff
@@ -292,17 +289,24 @@ private boolean hasCauseFromTheSamePullRequest(@Nullable List<Cause> causes, @Nu
-            logger.info(format("Build started (%s).", getBuilder().getProject().getName()));
-            this.stashPullRequestsBuilder.run();
+        if (stashPullRequestsBuilder == null) {
+            logger.info("Not ready to run.");
```
Shouldn't this be logged as error?

proski replied:
The documentation says it's possible. I understand it happens when using pipelines. I just don't know how bad it is. I think errors should be reserved for code logic errors and really bad misconfiguration.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.java:
```diff
@@ -292,17 +289,24 @@ private boolean hasCauseFromTheSamePullRequest(@Nullable List<Cause> causes, @Nu
+        }
+
+        stashPullRequestsBuilder.run();
+        getDescriptor().save();
```
It might be a noob question, but I don't understand why we need to save the descriptor at this point?

proski replied:
I have not idea, it was there and I didn't want to change it.

jakub-bochenski replied:
OK

proski committed:
[JENKINS-56349] Make StashPullRequestsBuilder constructor take two arguments

Call that constructor from StashBuildTrigger#start() instead of using
setters. Check that project is not null before passing it to the
StashPullRequestsBuilder constructor.

Remove setters from StashPullRequestsBuilder, they are not used anymore.
